src/main/java/pl/edu/mimuw/BankClient.java

```diff
@@ -1,35 +1,131 @@
 package pl.edu.mimuw;
 
+import java.util.ArrayList;
 import java.util.List;
 
 public class BankClient {
 
-  // TODO:
-  // - create proper constructor
-  // - implement methods
-  // - add your own methods (1 is enough)
+  private String name;
+  private String surname;
+  private int age;
+  private String accountNumber;
+  private List<BankAction> actionHistory;
+
+  public BankClient(
+    String name,
+    String surname,
+    int age,
+    String accountNumber
+  ) {
+    this.name = name;
+    this.surname = surname;
+    this.age = age;
+    this.accountNumber = accountNumber;
+    this.actionHistory = new ArrayList<>();
+  }
 
   public String getName() {
-    throw new IllegalStateException("TODO");
+    return name;
   }
 
   public String getSurname() {
-    throw new IllegalStateException("TODO");
+    return surname;
   }
 
   public int getAge() {
-    throw new IllegalStateException("TODO");
+    return age;
   }
 
   public String getAccountNumber() {
-    throw new IllegalStateException("TODO");
+    return accountNumber;
   }
 
+  /**
+   * @return zwraca kopie listy akcji wypelniona kopiami tych akcji
+   */
   public List<BankAction> getActionHistory() {
-    throw new IllegalStateException("TODO");
+    if (actionHistory == null) {
+      return null;
+    }
+    List<BankAction> overview = new ArrayList<>();
+    for (int i = 0; i < actionHistory.size(); i++) {
+      overview.add(actionHistory.get(i).createCopy());
+    }
+    return actionHistory;
   }
 
+  public BankClient createCopy() {
+    return new BankClient(
+      this.name,
+      this.surname,
+      this.age,
+      this.accountNumber
+    );
+  }
+
+  /**
+   * splaca tyle dlugow klienta ile moze, jesli jakies zobowiazanie spadnie do zera, jest likwidowane z listy
+   * @param amount kwota przeznaczona na splate wszytkich dlugow
+   * @return zwraca pozostala kwote
+   */
+  public double repayDebt(double amount) {
+    int i = 0;
+    if (actionHistory.isEmpty()) {
+      return amount;
+    } else {
+      while (i < actionHistory.size()) {
+        var action = actionHistory.get(i);
+        if (action instanceof LoanBankAction) {
```

Comment:
nie chcemy raczje robić takich akcji z rzutowanaimi i instanceof - możesz dodać do akcji jakąś metodę, która byłaby cząścią np tej metody, tak że w zależności na jakim typie akcji byśmy tę metodę zawołali, to rózne by było zachowanie - to jest moim zdaniem do przemyślenia jeszcze raz